src/main/scala/com/fulcrumgenomics/umi/CorrectUmis.scala

```diff
@@ -0,0 +1,229 @@
+    *  */
+  private[umi] case class UmiMatch(matched: Boolean, umi: String, mismatches: Int)
+
+  /**
+    * Metrics for the correction of UMI sequences to a fixed set of known UMIs.
+    *
+    * @param umi the corrected UMI sequence (or all Ns for unmatched)
+    * @param total_matches the number of UMI sequences that matched/were corrected to this UMI
+    * @param perfect_matches the number of UMI sequences that were perfect matches to this UMI
+    * @param one_mismatch_matches the number of UMI sequences that matched with a single mismatch
+    * @param two_mismatch_matches the number of UMI sequences that matched with two mismatches
+    * @param other_matches the number of UMI sequences that matched with > 2 mismatches
+    * @param fraction_of_matches the fraction of all UMIs that matched or were corrected to this UMI
+    * @param representation the [[total_matches]] for this UMI divided by the mean [[total_matches]] for all UMIs
+    */
+  case class UmiCorrectionMetrics(umi: String,
+                                  var total_matches: Long = 0,
+                                  var perfect_matches: Long = 0,
+                                  var one_mismatch_matches: Long = 0,
+                                  var two_mismatch_matches: Long = 0,
+                                  var other_matches: Long = 0,
+                                  var fraction_of_matches: Double = 0,
+                                  var representation: Double = 0
+                                 ) extends Metric
+}
+
+@clp(group=ClpGroups.Umi, description=
+  """
+    |Corrects UMIs stored in BAM files when a set of fixed UMIs is in use.  If the set of UMIs used in
+    |an experiment is known and is a subset of the possible randomers of the same length, it is possible
+    |to error-correct UMIs prior to grouping reads by UMI.  This tool takes an input BAM with UMIs in a
+    |tag (RX by default) and set of known UMIs (either on the command line or in a file) and produces:
+    |  1. A new BAM with corrected UMIs in the same tag the UMIs were found in
+    |  2. Optionally a set of metrics about the representation of each UMI in the set
+    |  3. Optionally a second BAM file of reads whose UMIs could not be corrected within the specific parameters
+    |
+    |All of the fixed UMIs must be of he same length, and all UMIs in the BAM file must also have the same
+    |length.  Multiple UMIs that are concatenated with hyphens (e.g. AACCAGT-AGGTAGA) are split apart,
+    |corrected individually and then re-assembled.  A read is accepted only if all the UMIs can be corrected.
+    |
+    |Correction is controlled by two parameters:
+    |  1. --max-mismatches controls how many mismatches (no-calls are counted as mismatches) are tolerated
+    |         between a UMI as read and a fixed UMI.
+    |  2. --min-distance controls how many more mismatches the next best hit must have
+    |
```

Comment:
I think it's worth saying that these parameters apply to matching each UMI individually, which matters in the case of having multiple UMIs (ex. `AAA-GGG`).